File: TwoSum.py
'''
Edge cases:
[1,3,7,9,2] t=11
[1,2,3,4,5] t=25
[] t=3
[5] t=8
[5] t=5
[1,6] t=7
'''
def twoSum(lst, t):
  #number to find = target - list[p1]
  # [p1,p2] -> [3,4]
  p1 = 0
  p2 = p1 +1
  
  for p1 in range(len(lst)):
    numberToFind = t - lst[p1]
    for p2 in range(len(lst)):
      if lst[p2] == numberToFind:
        return [p1, p2]
  #if there is no solution
  return 

lst = [1,3,7,9,2]
target = 11
print(twoSum(lst, target))

def SumTw0(lst, t):
  #{numberToF, index}
  hashT = {}
  for p in range(len(lst)):
    hashtVal = hashT[lst[p]]
    if hashtVal >= 0:
      return [hashtVal[p], p]
    else:
      numberToF = t - lst[p]
      hashT[numberToF] = p

# print(SumTw0(lst, target))

def sumTwo3(lst, t):
  hasT = {}
  for i, num in enumerate(lst):
    numToFind = t - num
    if numToFind in hasT:
      return [hasT[num, i]]
    else: 
      hasT[num] = i

print(sumTwo3(lst, target))

File: TwoSum.js
//Questions to ask the interviewer:
// Are there duplicate numbers in the array
// Are all the numbers positive or can it be any negatives
//Will there allways be a solution available
//What do we return if there is no solution
//Can there be multiple pairs that add up to the value?

//Two pointer technique
const findTwoSum = (nums, target) => {
  //Brute force, if there is a solutions this will find it:
  for (let p1 = 0; p1 < nums.length; p1++) {
    console.log({ p1, value: nums[p1] });
    const numberToFind = target - nums[p1];
    console.log({ numberToFind });
    for (let p2 = p1 + 1; p2 < nums.length; p2++) {
      if (nums[p2] === numberToFind) {
        return [p1, p2];
      }
    }
  }
  // if there is no solution
  return null;
};
let lst = [1, 3, 7, 9, 2];
let target = 11;
console.log(findTwoSum(lst, target));

const findTwoSum2 = (lst, target) => {
  // {numberToFind: index}
  const numsMap = {}; // space O(n)
  for (let p = 0; p < lst.length; p++) {
    // time O(n)
    console.log({ p, value: lst[p] });
    const currentMapVal = numsMap[lst[p]];
    console.log({ currentMapVal });
    if (currentMapVal >= 0) {
      return [currentMapVal, p];
    } else {
      let numberToFind = target - lst[p];
      console.log({ numberToFind });
      numsMap[numberToFind] = p;
    }
  }
  return null;
};

console.log(findTwoSum2(lst, target));
